plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+	}
+	if f.WriteTimeout < config.Duration(time.Second) {
+		f.WriteTimeout = config.Duration(time.Second * 10)
+	}
+
+	f.acc = acc
+
+	server := f.createHTTPServer()
+
+	f.wg.Add(1)
+	go func() {
+		defer f.wg.Done()
+		if err := server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Serve failed - %s", err.Error())
+			}
+			close(f.close)
+		}
+	}()
+
+	f.Log.Infof("Listening on %s", f.listener.Addr().String())
+
+	return nil
+}
+
+func (f *Firehose) createHTTPServer() *http.Server {
+	return &http.Server{
+		Addr:         f.ServiceAddress,
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
+		f.listener.Close()
+	}
+	f.wg.Wait()
+}
+
+func (f *Firehose) Init() error {
+	tlsConf, err := f.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", f.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	f.tlsConf = tlsConf
+	f.listener = listener
+
+	return nil
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
+	default:
+	}
+
+	if !choice.Contains(req.URL.Path, f.Paths) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r := &firehoseRequest{req: req}
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+	}
+
+	if err := r.validate(); err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	decodedBytesData, ok := r.decodeData()
+	if !ok {
+		f.Log.Error(formatLog(r.body.RequestID, "Failed to base64 decode record data"))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, bytes := range decodedBytesData {
+		m, err := f.Parse(bytes)
+		if err != nil {
+			f.Log.Error(formatLog(r.body.RequestID, "Unable to parse data"))
+			r.responseStatusCode = http.StatusBadRequest
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		once.Do(func() {
+			f.Log.Info(formatLog(r.body.RequestID, internal.NoMetricsCreatedMsg))
+		})
+	}
```

Comment:
Yeah I guess so. Since no metrics created we can return. I'll add it. Thanks for pointing out. Again just copied over this block from `http_listener_v2` plugin